### ISSUE TITLE
Join paths don't work when embeddable keys or domains on keys are present

### ISSUE THREAD

**Issue by NorbertKawinski** [jOOQ/jOOQ#16036]
### Expected behavior

Generate such query:
```
select "mcve"."book"."id", "mcve"."book"."title", "mcve"."book"."author_id"
from (
         "mcve"."book"
             join (
             "mcve"."author" as "alias_34803752"
                 join "mcve"."country" as "alias_31224187"
             on "alias_34803752"."country_id" = "alias_31224187"."id"
             )
         on "mcve"."book"."author_id" = "alias_34803752"."id"
         )
where "alias_31224187"."id" = '00000000-0000-0000-0000-000000000001'
```

### Actual behavior

Generates such query:
```
select "mcve"."book"."id", "mcve"."book"."title", "mcve"."book"."author_id"
from (
         "mcve"."book"
             join (
             "mcve"."author" as "alias_34803752"
                 join "mcve"."country" as "alias_31224187"
             on ("alias_34803752"."country_id") = (null)
             )
         on "mcve"."book"."author_id" = "alias_34803752"."id"
         )
where ("alias_31224187"."id") = ('00000000-0000-0000-0000-000000000001')
```

### Steps to reproduce the problem

Enable embeddables generation:
```
<embeddableDomains>.*</embeddableDomains>
```

Create following schema:
```
CREATE DOMAIN mcve.country_id AS uuid;
CREATE TABLE mcve.country
(
    id mcve.country_id PRIMARY KEY,
    name varchar(255) NOT NULL
);
INSERT INTO mcve.country (id, name)
VALUES ('00000000-0000-0000-0000-000000000001', 'Poland');

CREATE TABLE mcve.author
(
    id uuid PRIMARY KEY,
    name varchar(255) NOT NULL,
    born date,
    country_id mcve.country_id NOT NULL,
    CONSTRAINT fk_country FOREIGN KEY (country_id) REFERENCES mcve.country (id)
);
INSERT INTO mcve.author (id, name, born, country_id)
VALUES ('00000000-0000-0000-0000-000000000001', 'Adam Mickiewicz', '1798-12-24', '00000000-0000-0000-0000-000000000001');

CREATE TABLE mcve.book
(
    id uuid PRIMARY KEY,
    title varchar(255) NOT NULL,
    author_id uuid NOT NULL,
    CONSTRAINT fk_author FOREIGN KEY (author_id) REFERENCES mcve.author (id)
);
INSERT INTO mcve.book (id, title, author_id)
VALUES ('00000000-0000-0000-0000-000000000002', 'Pan Tadeusz', '00000000-0000-0000-0000-000000000001');
INSERT INTO mcve.book (id, title, author_id)
VALUES ('00000000-0000-0000-0000-000000000003', 'Dziady', '00000000-0000-0000-0000-000000000001');
```

Run following code:
```
var COUNTRY = new CountryIdRecord(UUID.fromString("00000000-0000-0000-0000-000000000001"));
var results1 = ctx.selectFrom(BOOK)
        .where(BOOK.author().country().ID.eq(COUNTRY))
        .fetch();
assertEquals(2, results1.size());
```

### jOOQ Version

org.jooq.pro:jooq:3.19.1

### Database product and version

postgres:16.1-alpine Docker image

### Java Version

Openjdk 21.0.1

### OS Version

Microsoft Windows 11 Pro [10.0.22621 Build 22621]

### JDBC driver name and version (include name if unofficial driver)

org.postgresql:postgresql:42.6.0

**Comment by NorbertKawinski** [comment]
Please find the MCVE here: https://github.com/NorbertKawinski/jOOQ-mcve/tree/mcve-for-issue-16036

**Comment by NorbertKawinski** [comment]
Actually, this simple query also behaves the same way when using embeddables:
```
var results = ctx.select(BOOK.author().country().ID).from(BOOK).fetch();
```

**Comment by lukaseder** [comment]
Thanks a lot for your report and MCVE. I'll try to reproduce this soon, and will look into a fix.

**Comment by lukaseder** [comment]
I can reproduce an issue like this:

```java
    @Test
    public void testEmbeddableDomainsInKeysWithPaths() {
        EmbeddableDomainsInKeysRecord r1 = ctx.newRecord(EMBEDDABLE_DOMAINS_IN_KEYS);

        DYearRecord y = new DYearRecord(2000);

        r1.setId(y);
        r1.setV(1);
        assertEquals(1, r1.store());

        EmbeddableDomainsChildRecord r2 = ctx.newRecord(EMBEDDABLE_DOMAINS_CHILD);
        r2.setId(1);
        r2.setEmbeddableDomainsInKeys(y);
        assertEquals(1, r2.store());

        assertEquals(y,
            ctx.select(EMBEDDABLE_DOMAINS_CHILD.embeddableDomainsInKeys().ID)
               .from(EMBEDDABLE_DOMAINS_CHILD)
               .fetchOne()
               .value1());
    }
```

The test fails with:

```
java.lang.NullPointerException: Cannot invoke "org.jooq.Field.equal(org.jooq.Field)" because "f1" is null
	at org.jooq_3.20.0-SNAPSHOT.H2_2_2_220.debug(Unknown Source)
	at org.jooq.impl.JoinTable.onKey0(JoinTable.java:822)
	at org.jooq.impl.AbstractContext$JoinNode.joinTree(AbstractContext.java:1301)
	at org.jooq.impl.AbstractContext$JoinNode.joinTree(AbstractContext.java:1265)
	at org.jooq.impl.DefaultRenderContext.scopeEnd0(DefaultRenderContext.java:311)
	at org.jooq.impl.AbstractContext.scopeEnd(AbstractContext.java:913)
	at org.jooq.impl.SelectQueryImpl.accept0(SelectQueryImpl.java:2016)
	at org.jooq.impl.SelectQueryImpl.accept(SelectQueryImpl.java:1556)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:726)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:371)
	at org.jooq.impl.AbstractQuery.getSQL0(AbstractQuery.java:513)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:303)
	at org.jooq.impl.AbstractResultQuery.fetchLazy(AbstractResultQuery.java:301)
	at org.jooq.impl.AbstractResultQuery.fetchLazyNonAutoClosing(AbstractResultQuery.java:322)
	at org.jooq.impl.SelectImpl.fetchLazyNonAutoClosing(SelectImpl.java:3256)
	at org.jooq.impl.ResultQueryTrait.fetchOne(ResultQueryTrait.java:509)
	at org.jooq.codegen.test.embeddabledomains.EmbeddableDomainsTest.testEmbeddableDomainsInKeysWithPaths(EmbeddableDomainsTest.java:246)
```

This seems worse than what you reported, there's a NPE when constructing the SQL, not a `NULL` value when running it. Will check if this is the same issue.

**Comment by lukaseder** [comment]
It probably is, it just seems to depend on whether the left hand side or the right hand side of the `=` predicate is `null` in the `onKey0` utility:

```java
    static final Condition onKey0(ForeignKey<?, ?> key, Table<?> fk, Table<?> pk) {
        Condition result = noCondition();

        TableField<?, ?>[] references = key.getFieldsArray();
        TableField<?, ?>[] referenced = key.getKeyFieldsArray();

        for (int i = 0; i < references.length; i++) {
            Field f1 = fk.field(references[i]);
            Field f2 = pk.field(referenced[i]);

            result = result.and(f1.equal(f2)); // Behaviour depends on whether f1 or f2 is null
        }

        return result;
    }
```

**Comment by lukaseder** [comment]
The problem is that `key` contains references to the embeddable's underlying `Field` instances, not to the embeddables. But the `Fields.field(Field)` lookup doesn't access those underlying fields. I'm not sure whether it should (for all users of the API), given that `<embeddableDomains/>` enables replacement of the underlying field by the embeddable.

If the `Fields.field(Field)` method behaviour isn't changed, we can still introduce internal API for this lookup.

**Comment by lukaseder** [comment]
Alternatively, `ForeignKey.getFieldsArray()` and `ForeignKey.getKeyFieldsArray()` need to be patched. That seems cleaner, because if the fields get replaced by the embeddables, then they shouldn't appear anymore in such meta API queries.

Both are breaking changes, though.

**Comment by lukaseder** [comment]
Related issue that was already fixed for `Table.getPrimaryKey()`:

- #15873

Reporter looks familiar :) The fix was along the lines of what I said here: https://github.com/jOOQ/jOOQ/issues/16036#issuecomment-1884628100, i.e. to let lookups return patched versions of `UniqueKey` and `ForeignKey`. #16036 only fixed an internal utility, with little effect on public API. We can probably do the same here.

**Comment by NorbertKawinski** [comment]
> Reporter looks familiar :) 
:)
Recently I had a deeper dive into the jOOQ.
I'm doing an experiment to create a projects without JPA/Hibernate.
So far I don't look back. It's a lot of fun and looks really promising thanks to updatable records, embeddables, implicit joins and many more great features. Thank You!

> with little effect on public API. We can probably do the same here.
That's good to hear :) Otherwise breaking changes would probably freeze this issue for a long time, right?
Do you often have decide whether to choose cleaner way, or the non-breaking way?

**Comment by NorbertKawinski** [comment]
Do you see any means of touching this issue anytime soon?

**Comment by lukaseder** [comment]
@NorbertKawinski Thanks for the ping. Let me have another look later this week.

**Comment by lukaseder** [comment]
Yeah, the fix was already clear with this comment:

- https://github.com/jOOQ/jOOQ/issues/16036#issuecomment-1884624651

There was an open question about whether there needs to be a more global fix for looking up fields that are replaced by embeddables (big change, risky), when fixing this particular issue in isolation would have been simple and not risky at all. Sorry for the delay.

This is now fixed for versions:

- 3.20.0
- 3.19.12 (https://github.com/jOOQ/jOOQ/issues/17151)
- 3.18.19 (https://github.com/jOOQ/jOOQ/issues/17152)
- 3.17.28 (https://github.com/jOOQ/jOOQ/issues/17153)

I realise, I hadn't answered your question:

> Do you often have to decide whether to choose cleaner way, or the non-breaking way?

The process is:

1. Try to find workarounds to get users unstuck. In this case, the workaround is just to avoid paths in this case. Though, I understand this is an annoying issue as you probably keep forgetting and running into this again and again.
2. Try to fix the immediate issue locally, which is usually non-risky.
3. Keep in mind the more general problem that might arise, and fix that maybe 3-5 years later, if it's not trivial

The 3rd bullet sounds like it takes a very long time, but it's always good to delay strategic decisions in a product. In this case, it became clear only way after embeddables were first implemented, that they're not so different from UDTs (where natively supported), nested records, and even tables themselves! There's a pending strategic issue about this topic, and it's not at all clear yet what it will mean:

- https://github.com/jOOQ/jOOQ/issues/10525

Looking up underlying fields from an embeddable field isn't so different from looking up `ROW` attributes, and guess what, I realised just now that we do have an issue for this!

- https://github.com/jOOQ/jOOQ/issues/16721

So, as you can see, it's never a good idea to rush these complex, strategic decisions. Otherwise, new isolated behaviours emerge rather than finding the bigger, more consistent picture.

In a nutshell, I recently tweeted this :)
https://x.com/lukaseder/status/1826229047472570769

**Comment by NorbertKawinski** [comment]
Thank you very much for taking care of this issue and even explaining the rationale behind you choices! <3
